### PR TITLE
[RHPAM-3640] Configure Kafka JBPM Emitter from Operator -7.11.x

### PR DESCRIPTION
**Pull request by desmax74**
See: https://issues.redhat.com/browse/RHPAM-3640
7.11.x
Signed-off-by: desmax74 <mdessi@redhat.com>

### CODE REVIEW

**Review by tchughesiv — CHANGES_REQUESTED** [review]
remove changes to any 7.11.0 files... including crd changes. this PR is targeting 7.11.1

**Comment by desmax74** [comment]
@spolti

**Comment by tchughesiv** [comment]
/assign @spolti 
@desmax74

**Comment by desmax74** [comment]
@jakubschwan  @spolti

**Review by tchughesiv — APPROVED** [review]

File: pkg/apis/app/v2/kieapp_types.go [hunk 1]
```diff
@@ -977,6 +979,20 @@ type KafkaExtObject struct {
...
+	// Jbpm emitter, a comma separated list of host/port pairs to use for establishing the initial connection to the Kafka cluster for Jbpm Emitter
+	BootstrapServers string `json:"bootstrapServers,omitempty"`
+	// Jbpm Emitter, identifier to pass to the server when making requests
+	ClientID string `json:"clientID,omitempty"`
+	// Jbpm Emitter, the number of acknowledgments the producer requires the leader to have received before considering a request complete.
+	Acks *int `json:"acks,omitempty"`
+	// Jbpm emitter, number of milliseconds that indicates how long publish method will bloc for jbpm emitter, default value is 2000
+	MaxBlockMs *int32 `json:"maxBlockMs,omitempty"`
+	// Date and time format to be sent to Kafka - default format is yyyy-MM-dd'T'HH:mm:ss.SSSZ
+	DateFormat string `json:"dateFormat,omitempty"`
```
Comment (spolti) [inline pkg/apis/app/v2/kieapp_types.go:993]: isn't this property missing? 
https://github.com/jboss-container-images/jboss-kie-modules/pull/456/files#diff-3e7628a0c0671f1c47f2f4da279754577748d791191318fb85c3d250b1ca182cR42

**Review by spolti — CHANGES_REQUESTED** [review]

File: pkg/controller/kieapp/defaults/defaults_test.go [hunk 2]
```diff
@@ -5951,3 +5951,64 @@ func TestCRServerCPULimitAndRequestUsingMilicores(t *testing.T) {
 	assert.Equal(t, values.Requests.Cpu(), env.Servers[0].DeploymentConfigs[0].Spec.Template.Spec.Containers[0].Resources.Requests.Cpu())
 	assert.Equal(t, values.Limits.Cpu(), env.Servers[0].DeploymentConfigs[0].Spec.Template.Spec.Containers[0].Resources.Limits.Cpu())
 }
+
+func TestRhdmEnvironmentWithKafkaJBPM(t *testing.T) {
```
Comment (spolti) [inline pkg/controller/kieapp/defaults/defaults_test.go:5955]: rhpam?

**Review by spolti — CHANGES_REQUESTED** [review]

File: pkg/apis/app/v2/kieapp_types.go (hunk 1 above)
Comment (desmax74) [reply pkg/apis/app/v2/kieapp_types.go:993]: No, we add in the common.yaml if the KafkaJbpmEventEmitters is present https://github.com/kiegroup/kie-cloud-operator/pull/654/files#diff-66b9e589dd9cc4038e91fe574a84c9d024cc6e65956366412f929a815e43cebaR946

File: pkg/controller/kieapp/defaults/defaults_test.go (hunk 2 above)
Comment (desmax74) [reply pkg/controller/kieapp/defaults/defaults_test.go:5955]: added

File: pkg/apis/app/v2/kieapp_types.go (hunk 1 above)
Comment (spolti) [reply pkg/apis/app/v2/kieapp_types.go:993]: ok, thx.

File: deploy/crds/kieapp.crd.yaml [hunk 3]
```diff
@@ -1913,6 +1913,35 @@ spec:
                                 type: string
                               type: array
                           type: object
+                        kafkaJbpmEventEmitters:
+                          description: KafkaJBPMEventEmittersObject kafka configuration
+                            to be used by the KieApp for JBPM Emitter
+                          properties:
+                            acks:
+                              description: Jbpm Emitter, the number of acknowledgments
```
Comment (spolti) [inline deploy/crds/kieapp.crd.yaml:1921]: please, align the descriptions here with the suggestions I made on the rhpam image.yaml, does not need to add the final part where it mentions what system property it sets.

**Review by spolti — CHANGES_REQUESTED** [review]
just one more thing. otherwise it is good to go.

File: deploy/crds/kieapp.crd.yaml (hunk 3 above)
Comment (desmax74) [reply deploy/crds/kieapp.crd.yaml:1921]: Done

File: pkg/apis/app/v2/kieapp_types.go [hunk 4]
```diff
@@ -977,6 +979,20 @@ type KafkaExtObject struct {
 	MaxBlockMs *int32 `json:"maxBlockMs,omitempty"`
 }
 
+// KafkaJBPMEventEmittersObject kafka configuration to be used by the KieApp for JBPM Emitter
```
Comment (spolti) [inline pkg/apis/app/v2/kieapp_types.go:982]:
```suggestion
// KafkaJBPMEventEmittersObject kafka configuration to be used by the KieApp for jBPM Emitter
```

File: deploy/ui/form.json [hunk 5]
```diff
@@ -1245,6 +1245,55 @@
...
+                  "jsonPath": "$.spec.objects.servers[*].kafkaJbpmEventEmitters",
+                  "visible": true,
+                  "fields": [
+                    {
+                      "label": "Bootstrap Servers",
+                      "type": "text",
+                      "required": false,
+                      "jsonPath": "$.spec.objects.servers[*].kafkaJbpmEventEmitters.bootstrapServers",
+                      "originalJsonPath": "$.spec.objects.servers[*].kafkaJbpmEventEmitters.bootstrapServers",
+                      "description": "A comma separated list of host/port pairs to use for establishing the initial connection to the Kafka cluster"
```
Comment (spolti) [inline deploy/ui/form.json:1261]: It sill not aligned with https://github.com/jboss-container-images/rhpam-7-openshift-image/pull/550/files

File: pkg/apis/app/v2/kieapp_types.go [hunk 6]
```diff
@@ -977,6 +979,20 @@ type KafkaExtObject struct {
...
+type KafkaJBPMEventEmittersObject struct {
+	// Jbpm emitter, comma separated list of host/port pairs to use for establishing the initial connection to the Kafka cluster.
```
Comment (spolti) [inline pkg/apis/app/v2/kieapp_types.go:984]: remove the jbpm emitter string from its description.
And seems it still not aligned with https://github.com/jboss-container-images/rhpam-7-openshift-image/pull/550/files

File: deploy/crds/kieapp.crd.yaml (hunk 3 above)
Comment (spolti) [reply deploy/crds/kieapp.crd.yaml:1921]: didn't see the changes here, the description here https://github.com/jboss-container-images/rhpam-7-openshift-image/pull/550/files is not correct, but here still not updated.

**Review by spolti — CHANGES_REQUESTED** [review]

File: pkg/apis/app/v2/kieapp_types.go (hunk 6 above)
Comment (desmax74) [reply pkg/apis/app/v2/kieapp_types.go]: ok

File: deploy/ui/form.json (hunk 5 above)
Comment (desmax74) [reply deploy/ui/form.json]: Updated

File: pkg/apis/app/v2/kieapp_types.go (hunk 4 above)
Comment (desmax74) [reply pkg/apis/app/v2/kieapp_types.go]: Updated

File: config/7.11.1/common.yaml [hunk 7]
```diff
@@ -941,6 +941,44 @@ servers:
...
+                    - name: KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_PROCESSES_TOPIC_NAME
+                      value: "[[.KafkaJbpmEventEmitters.ProcessesTopicName]]"
+                      #[[end]]
+                      #[[if .KafkaJbpmEventEmitters.TaskTopicName]]
+                    - name: KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_TASK_TOPIC_NAME
+                      value: "[[.KafkaJbpmEventEmitters.TaskTopicName]]"
+                      #[[end]]
+                      #[[if .KafkaJbpmEventEmitters.CasesTopicName]]
+                    - name: KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_CASES_TOPIC_NAME
+                      value: "[[.KafkaJbpmEventEmitters.CasesTopicName]]"
```
Comment (jakubschwan) [inline config/7.11.1/common.yaml:978]: Use plural in all topic names
```suggestion
                    - name: KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_PROCESSES_TOPIC_NAME
                      value: "[[.KafkaJbpmEventEmitters.ProcessesTopicName]]"
                      #[[end]]
                      #[[if .KafkaJbpmEventEmitters.TasksTopicName]]
                    - name: KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_TASKS_TOPIC_NAME
                      value: "[[.KafkaJbpmEventEmitters.TasksTopicName]]"
                      #[[end]]
                      #[[if .KafkaJbpmEventEmitters.CasesTopicName]]
                    - name: KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_CASES_TOPIC_NAME
                      value: "[[.KafkaJbpmEventEmitters.CasesTopicName]]"
```

File: deploy/crds/kieapp.crd.yaml [hunk 8]
```diff
@@ -1913,6 +1913,48 @@ spec:
...
+                                considering a request to be complete, not set by default.
+                              type: integer
+                            bootstrapServers:
+                              description: Comma separated list of host/port pairs
+                                to use for establishing the initial connection to
+                                the Kafka cluster.
+                              type: string
+                            casesTopicName:
+                              description: Name to override the default Case topic
+                                name
```
Comment (jakubschwan) [inline deploy/crds/kieapp.crd.yaml:1932]: Improve topic name description, please apply to all topics.
```suggestion
                              description: The topic name for cases event messages. Set up to override the default value jbpm-cases-events. 
```

Comment (desmax74) [reply deploy/crds/kieapp.crd.yaml:1932]: ok

File: config/7.11.1/common.yaml (hunk 7 above)
Comment (desmax74) [reply config/7.11.1/common.yaml:978]: ok

File: pkg/apis/app/v2/kieapp_types.go [hunk 9]
```diff
@@ -977,6 +979,26 @@ type KafkaExtObject struct {
...
+	// The number of acknowledgments that the emitter requires the leader to have received before considering a request to be complete, not set by default.
+	Acks *int `json:"acks,omitempty"`
+	// Value in milliseconds that indicates how long the 'publish' method will block the operation. Default 2000 milliseconds (2 seconds).
+	MaxBlockMs *int32 `json:"maxBlockMs,omitempty"`
+	// Date and time format to be sent to Kafka. Default format is yyyy-MM-dd'T'HH:mm:ss.SSSZ
+	DateFormat string `json:"dateFormat,omitempty"`
+	// Name to override the default Processes topic name
+	ProcessesTopicName string `json:"processesTopicName,omitempty"`
+	// Name to override the default Task topic name
+	TaskTopicName string `json:"taskTopicName,omitempty"`
```
Comment (spolti) [inline pkg/apis/app/v2/kieapp_types.go:997]: TasksTopicName?

**Review by spolti — CHANGES_REQUESTED** [review]

File: pkg/apis/app/v2/kieapp_types.go (hunk 9 above)
Comment (desmax74) [reply pkg/apis/app/v2/kieapp_types.go]: Just changed

File: deploy/crds/kieapp.crd.yaml (hunk 3 above)
Comment (desmax74) [reply deploy/crds/kieapp.crd.yaml]: updated

File: pkg/controller/kieapp/defaults/defaults_test.go [hunk 10]
```diff
@@ -5951,3 +5951,80 @@ func TestCRServerCPULimitAndRequestUsingMilicores(t *testing.T) {
...
+	case "KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_DATE_FORMAT":
+		assert.Equal(t, env.Value, "dd-MM-yyyy'T'HH:mm:ss.SSSZ")
+
+	case "KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_TASKS_TOPIC_NAME":
+		assert.Equal(t, env.Value, "my-tasks-topic")
+
+	case "KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_CASE_TOPIC_NAME":
+		assert.Equal(t, env.Value, "my-case-topic")
+
+	case "KIE_SERVER_KAFKA_JBPM_EVENT_EMITTER_PROCESS_TOPIC_NAME":
```
Comment (spolti) [inline pkg/controller/kieapp/defaults/defaults_test.go:6013]: is it correct? shouldn't the envs being set?

**Review by spolti — CHANGES_REQUESTED** [review]

File: pkg/controller/kieapp/defaults/defaults_test.go (hunk 10 above)
Comment (desmax74) [reply pkg/controller/kieapp/defaults/defaults_test.go:6013]: Are set here https://github.com/kiegroup/kie-cloud-operator/pull/654/files#diff-20d788a31791c7601d275bf50c8e30cb790834f4c05ce64c4e3bdfe92a6aaebaR6025-R6027

**Review by spolti — APPROVED** [review]

**Pull request merged**